opte/src/engine/packet.rs

```diff
@@ -751,19 +751,98 @@ impl Packet<Initialized> {
             seg_offset = 0;
         }
 
-        let body = BodyInfo {
+        assert!(
+            self.state.len >= pkt_offset,
+            "{} > {}",
+            self.state.len,
+            pkt_offset,
+        );
+
+        let mut body = BodyInfo {
             pkt_offset,
             seg_index,
             seg_offset,
             len: self.state.len - pkt_offset,
         };
-
         let flow = InnerFlowId::from(&info.meta);
+
+        // Packet processing logic requires all headers to be in the leading
+        // segment. Detect if this is not the case and squash segments
+        // containing headers into one segment. This value represents the
+        // inclusive upper bound of the squash.
+        let squash_to = match (body.seg_index, body.seg_offset) {
+            // The body is in the first segment meaning all headers are also in
+            // the first segment. No squashing needed.
+            (0, _) => 0,
+
+            // The body starts at a zero offset in segment n. This means we need
+            // to squash all segments prior to n.
+            (n, 0) => n - 1,
+
+            // The body starts at a non-zero offset in segment n. This means we
+            // need to squash all segments up to and including n.
+            (n, _) => n,
+        };
+
+        // If the squash bound is zero, there is notthing left to do here, just
+        // return.
+        if squash_to == 0 {
+            return Ok(Packet {
+                avail: self.avail,
+                source: self.source,
+                // The new packet is taking ownership of the segments.
+                segs: core::mem::take(&mut self.segs),
+                state: Parsed {
+                    len: self.state.len,
+                    hdr_offsets: info.offsets,
+                    meta: info.meta,
+                    flow,
+                    body_csum: info.body_csum,
+                    body,
+                    body_modified: false,
+                },
+            });
+        }
+
+        // Calculate the body offset within the new squashed segment
+        for s in &self.segs[..squash_to] {
+            body.seg_offset += s.len;
+        }
```

Comment:
Should this be conditional on `body.seg_offset > 0`?

Reply:
Yup. Fixed